File: java/leetcode/FlattenTree.java
//Leetcode problem 114 Flatten Binary Tree to Linked List
//Solution written by Xuqiang Fang on 3 April 2018 
import java.util.LinkedList;
class TreeNode{
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x){val = x;}
}

class Solution{
    public void flatten(TreeNode root){
        if(root == null)
            return;
        LinkedList<TreeNode> list = new LinkedList<>();
        list = get(root); 
        TreeNode cur = list.get(0);
        for(int i=1; i<list.size(); i++){
            cur.left = null;
            cur.right = list.get(i); 
            cur = cur.right;
        }
    }

    private LinkedList<TreeNode> get(TreeNode cur){
        LinkedList<TreeNode> list = new LinkedList<>();
        if(cur != null){
            list.add(cur);
            list.addAll(get(cur.left));
            list.addAll(get(cur.right));
        }
        return list;
    }
}

public class FlattenTree{
    public static void main(String[] args){
        Solution s = new Solution();
        TreeNode root = new TreeNode(1);
        TreeNode  a = new TreeNode(2);
        TreeNode  b = new TreeNode(5);
        TreeNode  c = new TreeNode(3);
        TreeNode  d = new TreeNode(4);
        TreeNode  e = new TreeNode(6);
        root.left = a;
        root.right = b;
        a.left = c;
        a.right = d;
        b.right = e;
        s.flatten(root);
        print(root);
    }
    public static void print(TreeNode root){
        if(root != null){
            System.out.println(root.val);
            //print(root.left);
            print(root.right);
        }
        else
            return;
    }
}

File: python/leetcode/FlattenTree.py
'''
Leetcode problem 114 Flatten Binary Tree to Linked List
Solution written by Xuqiang Fang on 3 April, 2018
'''
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def flatten(self, root):
        if(root == None):
            exit
        l = self.getList(root)
        if(len(l) == 0):
            exit
        cur = l[0]
        for i in range(1, len(l)):
            cur.left = None
            cur.right = l[i]
            cur = cur.right

    def getList(self, root):
        l = []
        if(root != None):
            l.append(root)
            l += self.getList(root.left)
            l += self.getList(root.right)
        return l

root = TreeNode(1)
a = TreeNode(2)
b = TreeNode(5)
c = TreeNode(3)
d = TreeNode(4)
e = TreeNode(6)
root.left = a
root.right = b
a.left = c
a.right = d
b.right = e
s = Solution()
s.flatten(root)
while(root != None):
    print root.val
    root = root.right
